### PR TITLE
OAuth (Login with Google) sign-in

### PR DESCRIPTION
**Pull request by RMuskovets**
# Checklist:

- [x] Added specs for all new ruby code
- [x] Green rspec run
- [x] Green rubocop
- [x] PR review from teammate

# What?
Added support for authentication via Google's OAuth provider.

## Why?
Because it was required for a task... And this is a nice improvement for the users who don't want to make a password.

## How?
Followed this guide on Medium: https://medium.com/@adamlangsner/google-oauth-rails-5-using-devise-and-omniauth-1b7fa5f72c8e. In short:
- added the OmniAuth gem along with its Google provider library
- added support for OmniAuth in User model
- created a controller for OAuth callbacks
- added a "Login with Google" link to the Sign In view

## Testing?
Did you write tests? No.

## Screenshots (optional)
Screenshot from 2021-06-16 21-22-08

## Anything Else?
I'm not sure if it's actually needed to include the OAuth-provided `uid` in User model so I didn't do it... Also I'm not including the OAuth credentials here, they should be provided by the `GOOGLE_OAUTH_CLIENT_ID` and `GOOGLE_OAUTH_CLIENT_SECRET` environment variables.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -11,4 +13,10 @@ class User < ApplicationRecord
                     format: { with: VALID_EMAIL_REGEX },
                     uniqueness: { case_sensitive: false }
   validates :password, presence: true, length: { minimum: 6, maximum: 256 }
+
+  def self.from_google(email:, other_options:)
```
Comment (andrjuha01) [inline app/models/user.rb:17]: Fix naming 'other_options'

File: app/models/user.rb [hunk 2]
```diff
@@ -11,4 +13,10 @@ class User < ApplicationRecord
...
+    # create_with(avatar_url: avatar_url, first_name: first_name, last_name: last_name,
```
Comment (andrjuha01) [inline app/models/user.rb:18]: Remove unused code

File: app/models/user.rb [hunk 3]
```diff
@@ -2,7 +2,9 @@ class User < ApplicationRecord
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
-         :recoverable, :rememberable, :validatable, :confirmable
+         :recoverable, :rememberable, :validatable, :confirmable,
+         :omniauthable,
+         omniauth_providers: [:google_oauth2]
```
Comment (andrjuha01) [inline app/models/user.rb:7]: Squash two lines into one

File: app/models/user.rb (hunk 1 above)
Comment (andrjuha01) [reply app/models/user.rb:17]: Add specs

File: app/helpers/users_helper.rb [hunk 4]
```diff
@@ -1,7 +1,22 @@
 module UsersHelper
+  GOOGLE_AVATAR_STORAGE_REGEX = %r{https://lh\d\.googleusercontent\.com.+}
+
+  def google_avatar_url_for(user, size)
```
Comment (andrjuha01) [inline app/helpers/users_helper.rb:4]: If not used outside - use private to hide

File: app/helpers/users_helper.rb [hunk 5]
```diff
@@ -1,7 +1,22 @@
...
+    url = user.avatar_url.split('=')[0] # crop off parameters
+    "#{url}=s#{size}-c"
+  end
+
   def gravatar_for(user, size: 40)
```
Comment (andrjuha01) [inline app/helpers/users_helper.rb:9]: Should it remain public?
If not - move to private

**Review by KostyaKobel — APPROVED** [review]
LGTM

File: app/models/user.rb (hunk 1 above)
Comment (RMuskovets) [reply app/models/user.rb]: Added

File: app/helpers/users_helper.rb [hunk 6]
```diff
@@ -4,4 +4,21 @@ def gravatar_for(user, size: 40)
...
+    if user.avatar_url && user.avatar_url =~ (GOOGLE_AVATAR_STORAGE_REGEX)
+      image_tag(google_avatar_url_for(user, size), alt: user.first_name, class: 'gravatar')
+    else
+      gravatar_for(user, size: size)
+    end
+  end
+
+  private
+
+  GOOGLE_AVATAR_STORAGE_REGEX = %r{https://lh\d\.googleusercontent\.com.+}
```
Comment (andrjuha01) [inline app/helpers/users_helper.rb:18]: Put constant at the top of the file

File: app/models/user.rb [hunk 7]
```diff
@@ -11,4 +12,8 @@ class User < ApplicationRecord
                     format: { with: VALID_EMAIL_REGEX },
                     uniqueness: { case_sensitive: false }
   validates :password, presence: true, length: { minimum: 6, maximum: 256 }
+
+  def self.from_google(email:, params:)
+    create_with(**params).find_or_create_by!(email: email)
```
Comment (andrjuha01) [inline app/models/user.rb:17]: Can we just use find_or_create_by without create_with?

**Review by andrjuha01 — APPROVED** [review]

**Pull request merged**